ewatercycle/models/lisflood.py

```diff
@@ -0,0 +1,398 @@
+    'ET0Maps': {'name': 'PrefixET0', 'value': 'et0'},
+}
+
+
+class Lisflood(AbstractModel):
+    """eWaterCycle implementation of Lisflood hydrological model.
+
+    Attributes:
+        bmi (Bmi): Basic Modeling Interface object
+        work_dir (PathLike): Working directory for the model where it can read/write files
+
+    Example:
+        See examples/lisflood.ipynb in `ewatercycle repository <https://github.com/eWaterCycle/ewatercycle>`_
+    """
+
+    # unable to subclass with more specialized arguments so ignore type
+    def setup(self,  # type: ignore
+              forcing: Union[ForcingData, PathLike],
+              parameterset: LisfloodParameterSet,
+              work_dir: PathLike = None) -> Tuple[PathLike, PathLike]:
+        """Configure model run
+
+        If forcing is missing evaporation files (e0, es0, et0) then the :py:meth:`run_lisvap` function should be run.
+
+        1. Creates config file and config directory based on the forcing variables and time range
+        2. Start bmi container and store as :py:attr:`bmi`
+
+        Args:
+            forcing: a forcing directory or a forcing data object.
+            parameterset: LISFLOOD input files. Any included forcing data will be ignored.
+            work_dir: a working directory given by user or created for user.
+
+        Returns:
+            Path to config file and path to config directory
+        """
+        singularity_image = CFG['lisflood.singularity_image']
+        docker_image = CFG['lisflood.docker_image']
+        self._check_work_dir(work_dir)
+        self._check_forcing(forcing)
+        self.parameterset = parameterset
+
+        config_file = self._create_lisflood_config()
+
+        if CFG['container_engine'].lower() == 'singularity':
+            self.bmi = BmiClientSingularity(
+                image=singularity_image,
+                input_dirs=[
+                    str(parameterset.root),
+                    str(parameterset.mask.parent),
+                    str(self.forcing_dir)
+                ],
+                work_dir=str(self.work_dir),
+            )
+        elif CFG['container_engine'].lower() == 'docker':
+            self.bmi = BmiClientDocker(
+                image=docker_image,
+                image_port=55555,
+                input_dirs=[
+                    str(parameterset.root),
+                    str(parameterset.mask.parent),
+                    str(self.forcing_dir)
+                ],
+                work_dir=str(self.work_dir),
+            )
+        else:
+            raise ValueError(
+                f"Unknown container technology in CFG: {CFG['container_engine']}"
+            )
+        return Path(config_file), self.work_dir
+
+    def _check_work_dir(self, work_dir):
+        """Check and optionally create the work directory."""
+        self.work_dir = _generate_workdir(work_dir)
+
+    def _check_forcing(self, forcing):
+        """"Check forcing argument."""
```

Comment:
Can we make this docstring slightly more informative?

Reply:
Sure, it is revised.